### PR TITLE
Improve Notification API

### PR DESCRIPTION
**Pull request by Hixie**
- better documentation

- verify (at run time) that onNotification doesn't return null, and
  report copious helpful information if it does.

- add a toString/debugFillDescription convention to Notification.

- actually test Notification

### CODE REVIEW

File: packages/flutter/lib/src/widgets/notification_listener.dart [hunk 1]
```diff
@@ -2,37 +2,61 @@
...
   }
+
+  @override
+  String toString() {
+    List<String> description = <String>[];
+    debugFillDescription(description);
+    return '$runtimeType(${description.join(", ")})';
+  }
+
+  void debugFillDescription(List<String> description) { }
```
Comment (abarth) [inline packages/flutter/lib/src/widgets/notification_listener.dart]: `@protected`

Comment (abarth) [inline packages/flutter/lib/src/widgets/notification_listener.dart]: `@mustCallSuper`

**Comment by abarth** [comment]
LGTM

File: packages/flutter/lib/src/widgets/notification_listener.dart (hunk 1 above)
Comment (Hixie) [inline packages/flutter/lib/src/widgets/notification_listener.dart]: done. I ever cargo-culted some docs for it while I was at it.

**Comment by Hixie** [comment]
Will check in on green.

**Pull request merged**